File: Solutions/221. Maximal Square/221. Maximal Square.java
/**
 * 
 */
package github.com.ErnestL1n;

import java.util.Arrays;

/**
 * @author https://github.com/ErnestL1n
 *
 */
//Leetcode.221
public class Maximal_Square {

	/**
	 * @param args
	 */
	
	public static int maximalSquare(char[][] matrix) {
		if(matrix==null||matrix.length==0||matrix[0].length==0)return 0;
		int row=matrix.length;
		int column=matrix[0].length;
		int[][] temp=new int[row][column];
		for(int i=0;i<row;i++)
			for(int j=0;j<column;j++)
				temp[i][j]=Character.getNumericValue(matrix[i][j]);
		int edge=0;
		for(int i=0;i<row;i++) {
			for(int j=0;j<column;j++) {
				if(i>0 && j>0 && temp[i][j]>0)
					temp[i][j]=Math.min(temp[i-1][j-1], Math.min(temp[i][j-1], temp[i-1][j]))+1;
				edge=Math.max(edge, temp[i][j]);
			}
		}
		return edge*edge;
	}
	public static void main(String[] args) {
		char[][] matrix=new char[][] {{'1','0','1','0','0'},{'1','0','1','1','1'},{'1','1','1','1','1'},{'1','0','0','1','0'}};
		System.out.println("Given a 2D binary matrix filled with 0's and 1's :");
		for(int i=0;i<matrix.length;i++)
			System.out.println(Arrays.toString(matrix[i]));
		System.out.println("the largest square containing only 1's is ");
		System.out.println(maximalSquare(matrix));
	}

}

File: Solutions/221. Maximal Square/221. Maximal Square.go
func maximalSquare(m [][]byte) int {
    dp:=make([][]int,len(m))
    for i,_:=range(dp){dp[i]=make([]int,len(m[0]))}
    edge:=0
    for i:=0;i<len(m);i+=1{
        for j:=0;j<len(m[0]);j+=1{
            if m[i][j]=='1' && i>0 && j>0{
                dp[i][j]=min(dp[i-1][j-1],min(dp[i-1][j],dp[i][j-1]))+1
            }else{
                dp[i][j]=int(m[i][j]-'0')
            }
            edge=max(edge,dp[i][j])
        }
    }
    return edge*edge
}

func min(i,j int)int{
    if i<j{
        return i
    }
    return j
}
func max(i,j int)int{
    if i<j{
        return j
    }
    return i
}